Add breakers for protection against Rx outages

aub opened the pull request:
This adds the new breakers gem for doing circuit breaking against outages on backend services. The bulk of the work is in the configuration for the gem, which happens in an initializer. It configures a service for Rx, which matches POST requests against the host and path defined in its configuration class. Once that configuration is in place, I just added the breakers middleware to the Faraday connection used by the Rx client. I've also added a test to show the exception being thrown when there is an outage against the service. I think this ought to be sufficient, since the breakers gem is itself well tested against various conditions, but I'm curious if people think there are other things that need testing.

For reviewers, I have a question about the Redis connection created in the initializer. Redis has its own initializer (redis.rb) which creates a connection and saves it in `Redis.current`. I would like to use that and not create a new one, but since the initializers are run in alphabetical order it will not have been initialized yet. I think the options are to do it this way or to prefix the initializer files with 01 and 02 to enforce the ordering. Although there is some code duplication involved, I think this way seems cleaner, since creating a Redis connection is so lightweight, but I'm interested in feedback there.

robbiethegeek commented:
This LGTM but would like to have @kreek and @saneshark give a quick review.

saneshark commented on config/initializers/breakers.rb:
```diff
@@ -0,0 +1,33 @@
+require 'rx/configuration'
+
+# Read the redis config, create a connection and a namespace for breakers
+redis_config = Rails.application.config_for(:redis).freeze
```
Why do think it is better to read this config here rather than using the REDIS_CONFIG constant that was already set in redis.rb? Oh... because of the order that they are loaded... hmmm

What if breakers was initialized from redis.rb after those constants were defined?

aub replied:
I think it's clearer for it to have its own initializer rather than sticking it into the Redis one, and the cost of creating another Redis connection seems low enough that it's fine.

saneshark commented on config/initializers/breakers.rb:
```diff
@@ -0,0 +1,33 @@
+  request_env.method == :post &&
+    request_env.url.host == rx_host &&
+    request_env.url.path =~ /^#{rx_path}/
+end
```
Are we only using breakers for POST request or for GET as well?

saneshark replied:
So I'm thinking as we add more Breakers, this initializer will quickly start to become unruly. Is it worth maybe having these within the Rx::Configuration singleton instead? And other services would follow a similar pattern?

aub replied:
You know, I'm not really sure what I was thinking about only using it for POST... good catch.

I'm working on the Secure Messaging thing as a separate PR right now, and part of that will involves creating the services in the Configuration classes. It'll be a lot cleaner.

saneshark replied:
Also please take a look at my open PR for refactoring Rx and SM

aub replied:
@saneshark No reason to rush. The only real change I'm making to the Rx and SM Client classes is the addition of the middleware, so it shouldn't be any problem to merge.

saneshark commented on lib/rx/client.rb:
```diff
@@ -91,7 +91,10 @@ def raise_not_authenticated
-      @connection ||= Faraday.new(config.base_path, headers: BASE_REQUEST_HEADERS, request: request_options)
+      @connection ||= Faraday.new(config.base_path, headers: BASE_REQUEST_HEADERS, request: request_options) do |conn|
+        conn.use :breakers
+        conn.adapter :httpclient
```
why does the adapter need to be httpclient instead of the default net/http?

aub replied:
It wasn't specifying an adapter before and needs to do so for this block configuration. I used httpclient because its what is being used by Sunsets, but I'll switch it back to the default one, as that's what's done in the secure messaging thing.

aub commented:
@saneshark Ok, I'll put together an integration test shortly.

aub commented:
I went ahead and put the SM work into this PR, since it's so similar to what's done for Rx. The latest commit addresses all of the issues described here other than the integration test, which I'm plugging away at now.

saneshark commented:
Reviewed changes, and LGTM.

kreek commented:
LGTM with more of a breakers question/todo: @robbiethegeek at one point we were going to have a meeting about metrics but I think that got lost in the shuffle; either way it would be cool if breakers could update CloudWatch as well as it's own dashboard (maybe via a metrics adapter if you want to keep it flexible for other projects to use StatsD or whatever).

robbiethegeek commented:
@kreek agreed about metrics, @plusjeff @jkassemi we should have a sync next week about this?  I am going to merge this in...

aub commented:
@kreek and @robbiethegeek I spoke with @jkassemi about this while working out the requirements, and it sounds like DevOps is setting up Prometheus and Graphene services that we can use for metrics collection and dashboards. I added a plugin system to Breakers to support that, so once those systems are set up it should be really easy to write a plugin that sends data to Prometheus.

saneshark commented:
@robbiethegeek I think @aub was still going to add an integration test. Guess that can go in a separate PR.

aub commented:
Yep, I'll make a separate PR.

robbiethegeek commented:
HAHA got excited for the LGTMs and merge happy :shipit:
